### PR TITLE
[maistra-2.4] OSSM-8001: Fix setting proxy GID

### PR DESCRIPTION
**Pull request by jewertow**
Our injection webhook had a bug in finding proxy GID, but it was hidden, because we were overwriting GID with UID in the injection template.
In general, we set UID=app_UID + 1 and GID=app_UID and we ignored explicitly set `runAsGroup`. This is incorrect, but users did not face any issue in restricted SCC, because there only `runAsUser` is set by OCP. In such a case, application GID is 0, so when we set proxyUID=appUID+1 and proxyGID=appUID we did not face any issues, because all IDs were unique.
This bug reveal when an application has explicitly set UID and GID, e.g. 0 and 0, because then the calculated app UID is 1 and app GID = 0. This means that GID of the app and proxy overlap, so the no-redirect rules for GID owner cause failures in traffic interception. In the last version of the Istio Operator we [fixed](https://github.com/maistra/istio-operator/pull/1787) the injection template and therefore that this long-living bug was exposed.

> [!IMPORTANT]
> Please review commits one by one, because then you will see how it worked (I added unit tests without changing logic in the first 2 commits) and how it works with the actual fixes.

### CODE REVIEW

File: pkg/kube/inject/webhook.go [hunk 1]
```diff
@@ -970,17 +970,23 @@ func findProxyUGID(pod corev1.Pod) (proxyUID, proxyGID *int64) {
...
 					if proxyUID == nil || uid > *proxyUID {
 						proxyUID = &uid
 					}
-					if proxyGID == nil {
-						proxyGID = c.SecurityContext.RunAsUser
+					gid := uid
+					if c.SecurityContext.RunAsGroup != nil {
+						gid = *c.SecurityContext.RunAsGroup + 1
+					}
+					if proxyGID == nil || gid > *proxyGID {
```
Comment (dgn) [inline pkg/kube/inject/webhook.go:988]: why do we check for `gid > *proxyGID`? Shouldn't the container's securityContext always take precedence, even if the int is lower?

**Review by dgn — APPROVED** [review]
lgtm, just one question

File: pkg/kube/inject/webhook.go (hunk 1 above)
Comment (jewertow) [reply pkg/kube/inject/webhook.go:988]: I am not the original author of this condition, so I do not know... But I would also expect that container SC takes precedence over pod SC.

Comment (luksa) [reply pkg/kube/inject/webhook.go:988]: All this code needs to do is find an UID and a GID that isn't used by any of the containers (i.e. it isn't specified in the pod's nor the container's `securityContext`).

But what the code effectively does is find the max UID/GID used by the containers and then adds a `+ 1`  to that value. Although this isn't wrong, it's not the best, as it wouldn't work if one of the containers uses the last value of the uid/gid range available in the pod's namespace.

So, ideally, we should fix this and return the first unused uid/gid we find. But I guess it's also okay if we leave the current behavior. It looks correct, apart from the corner case described above.

Comment (jewertow) [reply pkg/kube/inject/webhook.go:988]: > So, ideally, we should fix this and return the first unused uid/gid we find. But I guess it's also okay if we leave the current behavior. It looks correct, apart from the corner case described above.

@luksa do you suggest that my fix is incorrect? Now, this function finds first unused UID/GID, while previously it was only finding the first unused UID, but not GID.

Comment (luksa) [reply pkg/kube/inject/webhook.go:988]: No, I was just commenting on the `gid > *proxyGID` (which came from `uid > *proxyUID`). 

Your fix looks correct, but we should probably fix both uid and gid so that take the first unused UID/GID instead of taking `max(uid across all containers)+1`.

Comment (jewertow) [reply pkg/kube/inject/webhook.go:988]: Ok, thanks for clarifying.

Comment (dgn) [reply pkg/kube/inject/webhook.go:988]: we can consider a fix like that for upstream, but let's not overcomplicate the bugfix

**Comment by jtyler-knime** [comment]
@jewertow, very happy to see this! 

**Tl;Dr - This issue with group ID also surfaces when using the `nonroot-v2` SCC and setting `runAsUser` and `runAsGroup` to the same value.**

To give some additional context around this issue, we ran into this for a slightly different reason. 

In our use-case, we're distributing (via Helm) a chart that needs to be compatible with both OpenShift and other Kubernetes distributions for our end-users. 

Due to requirements in certain Kubernetes platforms, security context constraints **require** we explicitly set `runAsUser` and `runAsGroup` both to non-zero values. Our approach was to follow [Bitnami's lead](https://github.com/bitnami/charts/issues/24251) and set both user and group to the same value (i.e. `1001`; **and** subsequently requested users use the `nonroot-v2` SCC until we could update our chart(s) to be `restricted-v2` compatible). 

This worked fine for us with OSSM 2.5.2, but customer's reported issues when upgrading to OSSM 2.6 as this resulted in app containers and proxy containers having the same group id. 

Note that this wasn't immediately apparent to us, as the problem doesn't surface with upgrading OSSM **until** mesh-injected pods are actually created/recreated since that's when the injector web-hook fires. 

Any insight on when we might see this patched and available in 2.5 or 2.6 on Operator Hub?
